### PR TITLE
Re enable mods

### LINKED ISSUE #263: Mods that are disabled when joining a server are not re-enabled when leaving that server
This is annoying, mods that a server disables should re-enable when you leave the server, and not require you to manually enable them again.

Also on a similar note:
A popup appearing when you join a server that will disable a mod you have enabled would be nice. (I have lots of audio override mods so reload_mods takes like 20 seconds)

### PR DESCRIPTION
**Pull request by uniboi**
Enable installed mods that are required on a server and disable enabled mods that are specified as `RequiredOnClient` but not required for the server.
This closes #263
Also depends on [this](https://github.com/R2Northstar/NorthstarLauncher/pull/186) Launcher PR

Reload mods prompt:

Mods get reloaded after auth is complete:

### CODE REVIEW

**Comment by F1F7Y** [comment]
So uhhh, squirrel strings have a size limit. Maybe you should store `enabled_disabled_mods_pre_connect` in NorthstarLauncher instead of a ConVar :)

**Comment by uniboi** [comment]
(global) variables get reset because I'm reloading all mods.
I tried implementing reloading only specific mods but didn't finish it because another PR would change how this would behave.
Right now it essentially does what @F1F7Y suggested.

File: Northstar.Client/mod.json [hunk 1]
```diff
@@ -31,7 +31,11 @@
 		{
 			"Name": "filter_map_hide_locked",
 			"DefaultValue": "0"
-		}
+		},
+        {
```
Comment (F1F7Y) [inline Northstar.Client/mod.json:35]: funny tab length

File: Northstar.Client/mod/scripts/vscripts/ui/menu_ns_serverbrowser.nut [hunk 2]
```diff
@@ -226,6 +226,35 @@ void function InitServerBrowserMenu()
 
 	// UI was cut off on some aspect ratios; not perfect
 	UpdateServerInfoBasedOnRes()
+
+        // Listen for map unload to revert forced mod deactivation
+    AddUICallback_OnLevelInit( void function() // There's probably a better spot to put this into
```
Comment (F1F7Y) [inline Northstar.Client/mod/scripts/vscripts/ui/menu_ns_serverbrowser.nut:231]: why not put this in a separate func like most callbacks, just clutters up the already long serverbrowser init func

File: Northstar.Client/mod/resource/northstar_client_localisation_english.txt [hunk 3]
```diff
@@ -296,6 +296,12 @@ Press Yes if you agree to this. This choice can be changed in the mods menu at a
 		"INGAME_PLAYERS" "Players: ^6BA6C400%s1"
 		"TOTAL_SERVERS" "Servers: ^C46C6C00%s1"
 
+        "INFO_MOD_CHANGE" "Mods require reload"
+        "ENABLE_DISABLE_MODS_TO_CONTINUE" "Some mods need to be enabled or disabled before you can join this server"
+        "MODS_WILL_BE_DISABLED" "Mods that will be disabled (%i):"
+        "MODS_WILL_BE_ENABLED" "Mods that will be enabled (%i):"
+        "RELOAD_AND_CONTINUE" "Apply changes and continue"
+
```
Comment (RoyalBlue1) [inline Northstar.Client/mod/resource/northstar_client_localisation_english.txt:304]: Spaces when it should be tabs

File: Northstar.Client/mod/scripts/vscripts/ui/menu_ns_serverbrowser.nut [hunk 4]
```diff
@@ -1074,24 +1171,28 @@ void function ThreadedAuthAndConnectToServer( string password = "" )
...
+            modsChanged = true
+        }
+        foreach ( string mod in NSGetForcedDisabledMods() )
+        {
+            NSSetModEnabled( mod, false )
+            modsChanged = true
+        }
+
+        // only actually reload if we need to since the uiscript reset on reload lags hard
+        if ( modsChanged )
```
Comment (RoyalBlue1) [inline Northstar.Client/mod/scripts/vscripts/ui/menu_ns_serverbrowser.nut:1186]: Spaces when it should be tabs

File: Northstar.Client/mod/scripts/vscripts/ui/menu_ns_serverbrowser.nut [hunk 5]
```diff
@@ -1074,24 +1171,28 @@ void function ThreadedAuthAndConnectToServer( string password = "" )
...
+			// show a window so the user doesn't freak out that their game isn't responding for a few seconds
+			DialogData dialogData
+			dialogData.header = "Reloading Mods"
+			dialogData.image = $"ui/menu/common/dialog_error"
+			OpenDialog( dialogData )
 
-		// only actually reload if we need to since the uiscript reset on reload lags hard
-		if ( modsChanged )
-			ReloadMods()
+            ReloadMods()
```
Comment (RoyalBlue1) [inline Northstar.Client/mod/scripts/vscripts/ui/menu_ns_serverbrowser.nut:1194]: Spaces when it should be tabs

File: Northstar.Client/mod/resource/northstar_client_localisation_german.txt [hunk 6]
```diff
@@ -266,6 +266,12 @@ Drücke Ja, um zuzustimmen. Du kannst diese Entscheidung jederzeit im Modmenü 
 		"INGAME_PLAYERS" "Spieler: ^6BA6C400%s1"
 		"TOTAL_SERVERS" "Server: ^C46C6C00%s1"
 
+        "INFO_MOD_CHANGE" "Mods benötigen Änderungen"
+        "ENABLE_DISABLE_MODS_TO_CONTINUE" "Einige Mods müssen de/-aktiviert werden, bevor dem Server beigetreten werden kann"
+        "MODS_WILL_BE_DISABLED" "Mods die deaktiviert werden (%i):"
+        "MODS_WILL_BE_ENABLED" "Mods die aktiviert werden (%i):"
+        "RELOAD_AND_CONTINUE" "Änderungen übernehmen und fortfahren"
+
```
Comment (RoyalBlue1) [inline Northstar.Client/mod/resource/northstar_client_localisation_german.txt:274]: Spaces when it should be tabs

File: Northstar.Client/mod/scripts/vscripts/ui/menu_ns_serverbrowser.nut [hunk 7]
```diff
@@ -1027,9 +1061,72 @@ void function OnServerSelected( var button )
...
+        {
+            bool modRequired = requiredMods.contains( mod )
+            if ( NSIsModEnabled( mod ) != modRequired )
+                if ( modRequired )
+                    NSAddForcedEnabledMod( mod )
+                else
+                    NSAddForcedDisabledMod( mod )
+        }
+    }
+}
```
Comment (RoyalBlue1) [inline Northstar.Client/mod/scripts/vscripts/ui/menu_ns_serverbrowser.nut:1129]: Spaces when it should be tabs

**Review by x3Karma — APPROVED** [review]

**Comment by ASpoonPlaysGames** [comment]
Forgotten PR?

**Comment by uniboi** [comment]
Yes

**Pull request closed without merging**